File: lynx/lynx-pulsar/android/src/main/java/com/swmansion/pulsar/PulsarLynxModule.kt
package com.swmansion.pulsar

import android.content.Context
import com.lynx.jsbridge.LynxModule
import com.lynx.jsbridge.LynxMethod

/**
 * PulsarLynxModule — Lynx Native Module bridge for the Pulsar haptics SDK.
 *
 * Registration (in your Lynx host app):
 *   LynxEnv.inst().registerModule("PulsarModule", PulsarLynxModule::class.java)
 */
class PulsarLynxModule(context: Context) : LynxModule(context) {

    private val pulsar: Pulsar = Pulsar(context)
    private var realtimeComposer = pulsar.getRealtimeComposer()
    private var nextId: Int = 1
    private val patternComposersRegistry: MutableMap<Int, com.swmansion.pulsar.composers.PatternComposer> = mutableMapOf()

    // Pulsar -----------------------------------------------------------------

    @LynxMethod
    fun play(name: String) {
        pulsar.getPresets().getByName(name)?.play()
    }

    @LynxMethod
    fun enableHaptics(state: Boolean) {
        pulsar.enableHaptics(state)
    }

    @LynxMethod
    fun enableSound(state: Boolean) {
        pulsar.enableSound(state)
    }

    @LynxMethod
    fun enableCache(state: Boolean) {
        pulsar.enableCache(state)
    }

    @LynxMethod
    fun clearCache() {
        pulsar.clearCache()
    }

    @LynxMethod
    fun preloadPresets(names: List<String>) {
        pulsar.preloadPresets(names)
    }

    @LynxMethod
    fun stopHaptics() {
        pulsar.stopHaptics()
    }

    @LynxMethod
    fun shutDownEngine() {
        // no-op on Android
    }

    @LynxMethod
    fun hapticSupport(): Double {
        val support = pulsar.hapticSupport()
        return when (support) {
            com.swmansion.pulsar.types.CompatibilityMode.NO_SUPPORT -> 0.0
            com.swmansion.pulsar.types.CompatibilityMode.LIMITED_SUPPORT -> 2.0
            com.swmansion.pulsar.types.CompatibilityMode.STANDARD_SUPPORT -> 3.0
            else -> 4.0
        }
    }

    @LynxMethod
    fun forceHapticsSupportLevel(level: Double) {
        val mode = when (level.toInt()) {
            0 -> com.swmansion.pulsar.types.CompatibilityMode.NO_SUPPORT
            2 -> com.swmansion.pulsar.types.CompatibilityMode.LIMITED_SUPPORT
            3 -> com.swmansion.pulsar.types.CompatibilityMode.STANDARD_SUPPORT
            4 -> com.swmansion.pulsar.types.CompatibilityMode.ADVANCED_SUPPORT
            else -> com.swmansion.pulsar.types.CompatibilityMode.NO_SUPPORT
        }
        pulsar.forceHapticsSupportLevel(mode)
    }

    @LynxMethod
    fun enableImpulseCompositionMode(state: Boolean) {
        pulsar.enableImpulseCompositionMode(state)
    }

    @LynxMethod
    fun setRealtimeComposerStrategy(strategy: Double) {
        val strategyEnum = when (strategy.toInt()) {
            0 -> com.swmansion.pulsar.types.RealtimeComposerStrategy.ENVELOPE
            1 -> com.swmansion.pulsar.types.RealtimeComposerStrategy.PRIMITIVE_TICK
            2 -> com.swmansion.pulsar.types.RealtimeComposerStrategy.PRIMITIVE_COMPLEX
            3 -> com.swmansion.pulsar.types.RealtimeComposerStrategy.ENVELOPE_WITH_DISCRETE_PRIMITIVES
            else -> return
        }
        realtimeComposer = pulsar.getRealtimeComposer(strategyEnum)
    }

    // PatternComposer -----------------------------------------------------------------

    @LynxMethod
    fun parsePattern(data: Map<String, Any>): Int {
        val patternComposer = pulsar.getPatternComposer()

        val continuousMap = data["continuousPattern"] as? Map<String, Any>
        val discreteArray = data["discretePattern"] as? List<Map<String, Any>>

        val amplitudePoints = mutableListOf<com.swmansion.pulsar.types.ValuePoint>()
        (continuousMap?.get("amplitude") as? List<Map<String, Any>>)?.forEach { point ->
            amplitudePoints.add(
                com.swmansion.pulsar.types.ValuePoint(
                    time = (point["time"] as Number).toLong(),
                    value = (point["value"] as Number).toFloat()
                )
            )
        }

        val frequencyPoints = mutableListOf<com.swmansion.pulsar.types.ValuePoint>()
        (continuousMap?.get("frequency") as? List<Map<String, Any>>)?.forEach { point ->
            frequencyPoints.add(
                com.swmansion.pulsar.types.ValuePoint(
                    time = (point["time"] as Number).toLong(),
                    value = (point["value"] as Number).toFloat()
                )
            )
        }

        val continuousPattern = com.swmansion.pulsar.types.ContinuousPattern(
            amplitude = amplitudePoints,
            frequency = frequencyPoints
        )

        val discretePoints = mutableListOf<com.swmansion.pulsar.types.ConfigPoint>()
        discreteArray?.forEach { point ->
            discretePoints.add(
                com.swmansion.pulsar.types.ConfigPoint(
                    time = (point["time"] as Number).toLong(),
                    amplitude = (point["amplitude"] as Number).toFloat(),
                    frequency = (point["frequency"] as Number).toFloat()
                )
            )
        }

        val patternData = com.swmansion.pulsar.types.PatternData(
            continuousPattern = continuousPattern,
            discretePattern = discretePoints
        )
        patternComposer.parsePattern(patternData)

        val currentId = nextId
        nextId++
        patternComposersRegistry[currentId] = patternComposer
        return currentId
    }

    @LynxMethod
    fun patternPlay(patternId: Double) {
        patternComposersRegistry[patternId.toInt()]?.play()
    }

    @LynxMethod
    fun patternStop(patternId: Double) {
        patternComposersRegistry[patternId.toInt()]?.stop()
    }

    @LynxMethod
    fun patternRelease(patternId: Double) {
        patternComposersRegistry.remove(patternId.toInt())
    }

    // RealtimeComposer -----------------------------------------------------------------

    @LynxMethod
    fun realtimeSet(amplitude: Double, frequency: Double) {
        realtimeComposer.set(amplitude.toFloat(), frequency.toFloat())
    }

    @LynxMethod
    fun realtimePlayDiscrete(amplitude: Double, frequency: Double) {
        realtimeComposer.playDiscrete(amplitude.toFloat(), frequency.toFloat())
    }

    @LynxMethod
    fun realtimeStop() {
        realtimeComposer.stop()
    }

    @LynxMethod
    fun realtimeIsActive(): Boolean {
        return realtimeComposer.isActive()
    }
}

File: lynx/lynx-pulsar/ios/PulsarLynxModule.m
#import "PulsarLynxModule.h"
#import "Pulsar-Swift.h"

@implementation PulsarLynxModule {
  Pulsar *pulsar_;
  int nextId_;
  NSMutableDictionary<NSNumber *, PatternComposer *> *patternComposersRegistry_;
}

+ (NSString *)name {
  return @"PulsarModule";
}

+ (NSDictionary<NSString *, NSString *> *)methodLookup {
  return @{
    // Pulsar
    @"play":                          NSStringFromSelector(@selector(play:)),
    @"enableHaptics":                 NSStringFromSelector(@selector(enableHaptics:)),
    @"enableSound":                   NSStringFromSelector(@selector(enableSound:)),
    @"enableCache":                   NSStringFromSelector(@selector(enableCache:)),
    @"clearCache":                    NSStringFromSelector(@selector(clearCache)),
    @"preloadPresets":                NSStringFromSelector(@selector(preloadPresets:)),
    @"stopHaptics":                   NSStringFromSelector(@selector(stopHaptics)),
    @"shutDownEngine":                NSStringFromSelector(@selector(shutDownEngine)),
    @"hapticSupport":                 NSStringFromSelector(@selector(hapticSupport)),
    @"forceHapticsSupportLevel":      NSStringFromSelector(@selector(forceHapticsSupportLevel:)),
    @"enableImpulseCompositionMode":  NSStringFromSelector(@selector(enableImpulseCompositionMode:)),
    @"setRealtimeComposerStrategy":   NSStringFromSelector(@selector(setRealtimeComposerStrategy:)),

    // PatternComposer
    @"parsePattern":                  NSStringFromSelector(@selector(parsePattern:)),
    @"patternPlay":                   NSStringFromSelector(@selector(patternPlay:)),
    @"patternStop":                   NSStringFromSelector(@selector(patternStop:)),
    @"patternRelease":                NSStringFromSelector(@selector(patternRelease:)),

    // RealtimeComposer
    @"realtimeSet":                   NSStringFromSelector(@selector(realtimeSetAmplitude:frequency:)),
    @"realtimePlayDiscrete":          NSStringFromSelector(@selector(realtimePlayDiscreteAmplitude:frequency:)),
    @"realtimeStop":                  NSStringFromSelector(@selector(realtimeStop)),
    @"realtimeIsActive":              NSStringFromSelector(@selector(realtimeIsActive)),
  };
}

- (instancetype)init {
  self = [super init];
  if (self) {
    pulsar_ = [[Pulsar alloc] init];
    nextId_ = 1;
    patternComposersRegistry_ = [NSMutableDictionary new];
  }
  return self;
}

// Pulsar -----------------------------------------------------------------

- (void)play:(NSString *)name {
  [[[pulsar_ getPresets] getByName:name] play];
}

- (void)enableHaptics:(BOOL)state {
  [pulsar_ enableHapticsWithState:state];
}

- (void)enableSound:(BOOL)state {
  [pulsar_ enableSoundWithState:state];
}

- (void)enableCache:(BOOL)state {
  [pulsar_ enableCacheWithState:state];
}

- (void)clearCache {
  [pulsar_ clearCache];
}

- (void)preloadPresets:(NSArray *)presetNames {
  [pulsar_ preloadPresetsWithPresetNames:presetNames];
}

- (void)stopHaptics {
  [pulsar_ stopHaptics];
}

- (void)shutDownEngine {
  [pulsar_ shutDownEngine];
}

- (NSNumber *)hapticSupport {
  return [pulsar_ isHapticsSupported] ? @(4) : @(0);
}

- (void)forceHapticsSupportLevel:(double)level {
  // no-op on iOS
}

- (void)enableImpulseCompositionMode:(BOOL)state {
  // no-op on iOS
}

- (void)setRealtimeComposerStrategy:(double)strategy {
  // no-op on iOS
}

// PatternComposer -----------------------------------------------------------------

- (NSNumber *)parsePattern:(NSDictionary *)data {
  PatternComposer *patternComposer = [pulsar_ getPatternComposer];

  NSDictionary *continuousMap = data[@"continuousPattern"];
  NSArray *discreteArray = data[@"discretePattern"];

  NSMutableArray<ValuePoint *> *amplitudePoints = [NSMutableArray array];
  for (NSDictionary *point in continuousMap[@"amplitude"]) {
    ValuePoint *vp = [[ValuePoint alloc] initWithTime:[point[@"time"] doubleValue]
                                                value:[point[@"value"] floatValue]];
    [amplitudePoints addObject:vp];
  }

  NSMutableArray<ValuePoint *> *frequencyPoints = [NSMutableArray array];
  for (NSDictionary *point in continuousMap[@"frequency"]) {
    ValuePoint *vp = [[ValuePoint alloc] initWithTime:[point[@"time"] doubleValue]
                                                value:[point[@"value"] floatValue]];
    [frequencyPoints addObject:vp];
  }

  ContinuousPattern *continuousPattern =
    [[ContinuousPattern alloc] initWithAmplitude:amplitudePoints
                                       frequency:frequencyPoints];

  NSMutableArray<DiscretePoint *> *discretePoints = [NSMutableArray array];
  for (NSDictionary *point in discreteArray) {
    DiscretePoint *dp = [[DiscretePoint alloc] initWithTime:[point[@"time"] doubleValue]
                                                  amplitude:[point[@"amplitude"] floatValue]
                                                  frequency:[point[@"frequency"] floatValue]];
    [discretePoints addObject:dp];
  }

  PatternData *patternData =
    [[PatternData alloc] initWithContinuousPattern:continuousPattern
                                   discretePattern:discretePoints];
  [patternComposer parsePatternWithHapticsData:patternData];

  int currentId = nextId_;
  nextId_++;
  patternComposersRegistry_[@(currentId)] = patternComposer;
  return @(currentId);
}

- (void)patternPlay:(double)patternId {
  [patternComposersRegistry_[@((int)patternId)] play];
}

- (void)patternStop:(double)patternId {
  [patternComposersRegistry_[@((int)patternId)] stop];
}

- (void)patternRelease:(double)patternId {
  [patternComposersRegistry_ removeObjectForKey:@((int)patternId)];
}

// RealtimeComposer -----------------------------------------------------------------
// no-op on iOS — CoreHaptics realtime API requires different architecture

- (void)realtimeSetAmplitude:(double)amplitude frequency:(double)frequency {
  // no-op on iOS
}

- (void)realtimePlayDiscreteAmplitude:(double)amplitude frequency:(double)frequency {
  // no-op on iOS
}

- (void)realtimeStop {
  // no-op on iOS
}

- (NSNumber *)realtimeIsActive {
  return @(NO);
}

@end
